### PR TITLE
Best practice compliance recording

### PR DESCRIPTION
**Pull request by iawells**
Not expecting this to merge; Taylor and I are currently working on it.

### CODE REVIEW

**Comment by taylor** [comment]
Comments and feedback welcome

File: cbpps/0003-exceptions.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+- [ ] (R) CBPP approvers have approved the CBPP status as `implementable`
+- [ ] (R) CBPP summary, motivation and best practice details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to CNF Test Suite input
+- [ ] (R) Scoring has been determined
+- [ ]   "Implementation History" section is up-to-date
+- [ ]    Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## **Summary**
+
+CNFs should be compliant with applicable best practices, and ops processes
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:39]: Don't both the CNF and the operational processes have to be compliant and, thus, have to be validated for compliance with the best practices?

File: cbpps/0003-exceptions.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+should be validated for compliance with them.  However, compliance is a
+journey, and we expect end users to start with little compliance and
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:41]: I don't think it is "end users" compliance but the artifacts (CNFs, and associated processes including operational processes) that need to be compliant. 

Maybe this sentence, starting from the previous line should read (I don't believe "However" is appropriate):
"As compliance is a journey, we expect that the CNF's and operational processes will improve compliance over time, starting with limited or no compliance, progressing to a high degree and eventually full compliance."

File: cbpps/0003-exceptions.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+improve over time.
+
+Compliance is most useful when users, auditors and so on can identify
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md]: "... auditors and so ..." -- is there a typo here with "so"?

File: cbpps/0003-exceptions.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+exactly _how_ and _why_ you are compliant.  If they test your compliance,
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:45]: should there be a "what level" of compliance?

File: cbpps/0003-exceptions.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+and their tests flag problems, they can see that this is intentional, and
+they can find out why compliance is not possible.  They can see why you
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:47]: should "... find out why compliance is not possible" be changed to the "why compliance is limited to only a certain level"?

File: cbpps/0003-exceptions.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Workload Context](#workload-context)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:19]: Should this be not at a sub-level?
Also, should this be required? I strongly suggest that.

File: cbpps/0003-exceptions.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+Compliance is most useful when users, auditors and so on can identify
+exactly _how_ and _why_ you are compliant.  If they test your compliance,
+and their tests flag problems, they can see that this is intentional, and
+they can find out why compliance is not possible.  They can see why you
+chose not to be compliant and whether this non-compliance is acceptable to
+them.  They can also review any mitigating steps you might propose for this
+non-compliance or work out steps or their own.
+
+If your team builds a component used in NFV, or operates a component
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:52]: typo
```suggestion
If your team builds a component used in an NFV or operates a component
```

File: cbpps/0003-exceptions.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+of NFV, this document explains how you document your compliance
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:53]:
```suggestion
of an NFV, this document explains how you document your compliance
```

File: cbpps/0003-exceptions.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+Best practices do not always coincide with pragmatic design.  For whatever
+reason, a best practice may not be applicable to your circumstance; it
+may be too expensive to implement; it may even be technologically
+impossible based on other constraints.
+
+Our task as a working group is to help CNF users do _better_, by painting a
+picture of a perfect world and explaining how teams can progressively
+take steps towards it.  We want to help people on the journey.  And during
+that journey, their components and processes will be imperfect.
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:68]:
```suggestion
that journey, the components and processes may likely be imperfect.
```

File: cbpps/0003-exceptions.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+If your team builds a component used in NFV, or operates a component
+of NFV, this document explains how you document your compliance
+status.  We recommend you make this best practice a part of your
+delivery process.  This should be the very first best practice you attempt
+to comply with.
+
+## **Motivation**
+
+Best practices do not always coincide with pragmatic design.  For whatever
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:60]: The paragraph that follows is the "motivation" -- this current paragraph is a "qualifier" to the "imperfect" in the last sentence of the following paragraph. Hence, I suggest that we move this paragraph to be the second paragraph.

File: cbpps/0003-exceptions.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+By understanding that exceptions are going to exist, we make it possible
+for teams to understand their shortcomings and document them.  Where
+exceptions to specific best practices are frequent, this will also
+feed back into improvements to our best practice recommendations.
+
+### **Goals**
+
+* Let users explain which best practices they are compliant with
+* Communicate the level of compliance of individual parts of the system,
+  both software and process level
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:79]:
```suggestion
  both at the software and process level
```

File: cbpps/0003-exceptions.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+* Let users use CNF-WG best practices even if they cannot be compliant
+  with every single recommendation
+* Help users reduce the number of exceptions over time
+
+### **Non-Goals**
+
+* Providing a specific file format for the documentation
+* Requiring extensive documentation of each element - we will keep
+  required information to a minimum
+* Preventing delivery of components where non-compliance exists - we want
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:89]: Can we make this the #1 Non-Goal?

File: cbpps/0003-exceptions.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+  teams to be able to deliver components that are good enough, even if they
+  cannot be perfect
+* Documenting processes for improving compliance - we recommend teams
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:92]: Can we make this the #2 Non-Goal?

File: cbpps/0003-exceptions.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+* Help users reduce the number of exceptions over time
+
+### **Non-Goals**
+
+* Providing a specific file format for the documentation
+* Requiring extensive documentation of each element - we will keep
+  required information to a minimum
+* Preventing delivery of components where non-compliance exists - we want
+  teams to be able to deliver components that are good enough, even if they
+  cannot be perfect
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:91]:
```suggestion
  are not perfect
```

File: cbpps/0003-exceptions.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+* Documenting processes for improving compliance - we recommend teams
+  regularly review their compliance list, but we do not mandate any specific
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:93]:
```suggestion
  regularly review their compliance, but we do not mandate any specific
```

File: cbpps/0003-exceptions.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+  method and this will depend on the team and their working structure
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:94]: if we are not specifying then it is implicit that it depends on the teams

```suggestion
  method or periodicity of reviews
```

File: cbpps/0003-exceptions.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+
+### Format examples
+
+A simple machine-readable format might be a table in CSV format.  This
+can contain the records described, and is conveniently machine-readable.
+
+... example here ...
+
+Making machines conveniently machine-readable, such as records of compliance
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md]:
```suggestion
Making records conveniently machine-readable, such as records of compliance
```

File: cbpps/0003-exceptions.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+and non-compliance as simple booleans, helps consumers in filtering these
+records and even building compliance checks into their acceptance tests.
+Similarly, using a standard encoded format for the CNF-WG best practice
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:114]: "encoded" for "names"?
```suggestion
Similarly, using a standard format for the CNF-WG best practice
```

File: cbpps/README.md [hunk 19]
```diff
@@ -13,3 +13,20 @@ Follow the process outlined in the [CBPP process document](0001-cnf-best-practic
...
+
+CNFs should be compliant with applicable best practices, and ops processes
+should be validated for compliance with them.  However, compliance is a
+journey, and we expect end users to start with little compliance and
+improve over time.  The [CBPP-3](0003-exceptions.md) discusses how you should
+document your compliance and any places where you require an exception for
+being non-compliant.  This allows any auditor (e.g. a consumer of your CNF,
+your security group, an external auditing entity) to determine where you
+are not following best practices, _why_ you are not following best practices,
+whether you have mitigations in place for any risks this may cause, and
```
Comment (pgoyal01) [inline cbpps/README.md:27]: mitigations has not been covered elswhere

File: cbpps/0003-exceptions.md (hunk 7 above)
Comment (Smitholi67) [inline cbpps/0003-exceptions.md:52]: I do not understand why we are mentioning NFVs.  Perhaps this is a vocabulary issue but for me VNFs and CNFs are not the same.

Comment (pgoyal01) [reply cbpps/0003-exceptions.md:52]: @Smitholi67 NFV can be treated as a concept independent of the implementation

File: cbpps/0003-exceptions.md (hunk 2 above)
Comment (taylor) [reply cbpps/0003-exceptions.md:41]: @pgoyal01 do a suggest edit and tag @iawells please

File: cbpps/0003-exceptions.md (hunk 3 above)
Comment (taylor) [inline cbpps/0003-exceptions.md]:
```suggestion
Compliance is most useful when users, auditors, and other interested parties can identify
```

Comment (taylor) [reply cbpps/0003-exceptions.md]: @pgoyal01 I think that "so on" was communicating other parties similar to the users and auditors.  I have added a suggest edit along these lines.

File: cbpps/0003-exceptions.md (hunk 5 above)
Comment (taylor) [reply cbpps/0003-exceptions.md:47]: @pgoyal01 this is about compliance with a specific single best practice.  It is not communicating how close or not close someone is to being compliant.

File: cbpps/0003-exceptions.md (hunk 3 above)
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:44]: @iawells @taylor 
```suggestion
Compliance is most useful when users, auditors, and others can identify
```

Comment (pgoyal01) [reply cbpps/0003-exceptions.md]: @taylor that is great

File: cbpps/0003-exceptions.md (hunk 4 above)
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:45]: @taylor @iawells 
```suggestion
exactly _how_ , _why_ and at _what level_ you are compliant.  If they test your compliance,
```

File: cbpps/0003-exceptions.md (hunk 5 above)
Comment (pgoyal01) [reply cbpps/0003-exceptions.md:47]: @taylor Thanks for the clarification. The best practice is, IMHO, non-atomic and hence my question.

File: cbpps/0003-exceptions.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+## **Release Signoff Checklist**
+
+Items marked with (R) are required for the proposed best practice to be included in a release.
+
+- [ ] (R) CBPP approvers have approved the CBPP status as `implementable`
+- [ ] (R) CBPP summary, motivation and best practice details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to CNF Test Suite input
+- [ ] (R) Scoring has been determined
+- [ ]   "Implementation History" section is up-to-date
+- [ ]    Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
```
Comment (electrocucaracha) [inline cbpps/0003-exceptions.md]: nit
```suggestion
- [ ]    Supporting documentation — e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
```

File: cbpps/0003-exceptions.md (hunk 6 above)
Comment (taylor) [reply cbpps/0003-exceptions.md:19]: @pgoyal01 make a suggest edit

File: cbpps/0003-exceptions.md (hunk 14 above)
Comment (taylor) [reply cbpps/0003-exceptions.md:91]: @pgoyal01 there can be situations where something is not able to follow a best practice. That is different from not yet following a best practice

File: cbpps/0003-exceptions.md (hunk 6 above)
Comment (pgoyal01) [reply cbpps/0003-exceptions.md:19]:
```suggestion
    - [Notes/Constraints/Caveats (Mandatory)](#notesconstraintscaveats-mandatory)
```

File: cbpps/0003-exceptions.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+
+## **Summary**
+
+CNFs should be compliant with applicable best practices, and ops processes
+should be validated for compliance with them.  However, compliance is a
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:40]: @iawells @taylor 
```suggestion
should be validated for compliance with them.  As compliance is a
```

File: cbpps/0003-exceptions.md (hunk 2 above)
Comment (pgoyal01) [reply cbpps/0003-exceptions.md:41]: @iawells @taylor 
```suggestion
journey, we expect that the CNFs and operational processes will
```

File: cbpps/0003-exceptions.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+journey, and we expect end users to start with little compliance and
+improve over time.
```
Comment (pgoyal01) [inline cbpps/0003-exceptions.md:42]: @iawells @taylor 
```suggestion
improve compliance over time, starting with limited or no compliance, progressing to a high degree, and eventually full compliance.
```

File: cbpps/0003-exceptions.md (hunk 14 above)
Comment (pgoyal01) [reply cbpps/0003-exceptions.md:91]: @taylor Thanks. Agre "are not" is incorrect in that situation. It is just "cannot" is associated with a follow-up "why not" and hence was looking for an alternate. We can resolve this conversation, for some reason I don't see an ability to do so.

**Comment by electrocucaracha** [comment]
@iawells can you review it again and accept some comments?

File: cbpps/0003-exceptions.md (hunk 6 above)
Comment (electrocucaracha) [reply cbpps/0003-exceptions.md:19]: This suggestion makes more sense to change in the original template. Open a new PR in other issue.

File: cbpps/0003-exceptions.md (hunk 12 above)
Comment (electrocucaracha) [reply cbpps/0003-exceptions.md:89]: This is a non-ordered list

File: cbpps/0003-exceptions.md (hunk 13 above)
Comment (electrocucaracha) [reply cbpps/0003-exceptions.md:92]: ditto line 89

**Review by electrocucaracha — APPROVED** [review]

**Review by taylor — APPROVED** [review]
lgtm

**Comment by rich-l** [comment]
Looks good to me.

**Comment by agentpoyo** [comment]
:+1:

**Comment by wvwatson** [comment]
👍

**Pull request merged**